src/celeritas/optical/CerenkovPreGenerator.hh

```diff
@@ -0,0 +1,139 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/optical/CerenkovPreGenerator.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "corecel/Assert.hh"
+#include "corecel/Macros.hh"
+#include "celeritas/phys/ParticleTrackView.hh"
+#include "celeritas/random/distribution/PoissonDistribution.hh"
+#include "celeritas/track/SimTrackView.hh"
+
+#include "CerenkovData.hh"
+#include "CerenkovDndxCalculator.hh"
+#include "OpticalDistributionData.hh"
+#include "OpticalPropertyData.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Sample the number of Cerenkov photons to be generated by
+ * \c CerenkovGenerator and populate \c OpticalDistributionData values using
+ * Param and State data.
+ * \code
+    CerenkovPreGenerator::InputStepStateData step_data;
+    // Populate step_data using ParticleTrackView, SimTrackView, etc
+
+    CerenkovPreGenerator pre_cerenkov(
+        properties, cerenkov_data, step_data);
+
+    auto optical_dist_data = pre_cerenkov(rng);
+    if (optical_dist_data)
+    {
+        CerenkovGenerator cerenkov_gen(... , optical_dist_data, ...);
+        cerenkov_gen(rng);
+    }
+ * \endcode
+ */
+class CerenkovPreGenerator
+{
+  public:
+    struct InputStepStateData
+    {
+        real_type time{};  //!< Pre-step time
+        real_type step_length{};  //!< Step length
+        units::ElementaryCharge charge;
+        OpticalMaterialId material;
+        EnumArray<StepPoint, OpticalStepData> points;  //!< Pre- and post-steps
```

Comment:
This may be a reasonable placeholder; however, I think what you had originally is closer to what this will ultimately look like. Since this input is essentially the same as `OpticalDistributionData`, now we would be creating the distribution outside the pre-generator, passing it in, copying it into a new struct and getting it back (plus the sampled number of photons). If we were to do things this way it would make more sense to have this class just sample and return the number of photons.

Reply:
Fair. If we go for constructing it directly with views (so `ParticleTrack`, `SimTrack`, `GeoTrack` views), we'd still be missing pre-step time, right?

Reply:
Right, and pre-step position and speed. I think your extra struct with the data we can't get directly from the track views was a good solution for now (maybe `OpticalPreStepData` instead of `OpticalStepCollectorData`?). We can assume we'll have that data and all the track views when we construct the pre-generator.

Reply:
Okay. Refactored so all the data that can come from views is pulled directly from them. Unavailable data is cached in a temporary placeholder, as you suggested.